### PR TITLE
proc_mgr: introducing the subscription API

### PR DESCRIPTION
**Pull request by uLipe**
to receive data from nodes that completed their
processing chain.

Signed-off-by: Felipe Neves <felipe.neves@linaro.org>

### CODE REVIEW

**Comment by uLipe** [comment]
@microbuilder PTAL

**Comment by uLipe** [comment]
altough it has sample, I think this needs a proper unit tests before get merged.

**Comment by microbuilder** [comment]
> altough it has sample, I think this needs a proper unit tests before get merged.

I was just going to say that. It really should have a test suite, but I'll do some testing of the sample, thanks!

File: Kconfig.step [hunk 1]
```diff
@@ -108,4 +108,11 @@ config STEP_PROC_MGR_STACK_SIZE
 	  account the requirements of the largest processor node chain in the
 	  pipeline.
 
+config STEP_PROC_MGR_CALLBACKS_NUM
+	int "Maximum number of callbacks available to allocate"
+	default 32
+	help
+	  Sets the maximum number of callback that the processor manager
```
Comment (microbuilder) [inline Kconfig.step:115]:
```suggestion
	  Sets the maximum number of callbacks that the processor manager
```

File: include/step/proc_mgr.h [hunk 2]
```diff
@@ -44,6 +44,14 @@
 #ifdef __cplusplus
 extern "C" {
 #endif
+/**
+ * @brief Calllback deferred when a node chain is completed.
```
Comment (microbuilder) [inline include/step/proc_mgr.h:48]:
```suggestion
 * @brief Callback fired when a node chain is completed.
```

File: samples/simple_subscription/README.rst [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+Secure Telemetry Pipeline (STeP) Subscription Example
+#####################################################
+
+Overview
+********
+
+This sample exposes a simple subscription procedure, a STeP node chain
+is created then the external application subscribe to its completion and
+republishes the data to the chain.
```
Comment (microbuilder) [inline samples/simple_subscription/README.rst:11]:
```suggestion
This sample demonstrates how to use the subscription API.
A STeP node chain is created, and the external application
subscribes to the node chain's completion event, republishing
the data to the node chain.
```

File: src/proc_mgr.c [hunk 4]
```diff
@@ -26,6 +44,11 @@ struct step_pm_node_record {
 	 */
 	sys_snode_t snode;
 
+	/**
+	 * @brief Singly-linked list of subscribers callback.
```
Comment (microbuilder) [inline src/proc_mgr.c:48]:
```suggestion
	 * @brief Singly-linked list of subscriber callbacks.
```

**Review by microbuilder — CHANGES_REQUESTED** [review]

**Comment by uLipe** [comment]
Now we have tests for that. 

@microbuilder PTAL

**Review by microbuilder — APPROVED** [review]

**Pull request merged**